routemaster/state_machine/utils.py

```diff
@@ -131,11 +124,41 @@ def lock_label(label: LabelRef, conn):
     return row
 
 
-def labels_to_retry_for_action(
+def labels_in_state(
+    state_machine: StateMachine,
+    state: State,
+    conn,
+) -> Iterable[str]:
+    """Util to get all the labels in an action state that need retrying."""
+    ranked_transitions = select((
+        history.c.label_name,
+        history.c.old_state,
+        history.c.new_state,
+        func.row_number().over(
+            order_by=history.c.created.desc(),
+            partition_by=history.c.label_name,
+        ).label('rank'),
+    )).where(
+        history.c.label_state_machine == state_machine.name,
+    ).alias('transitions')
+
+    active_participants = select((
+        ranked_transitions.c.label_name,
+    )).where(and_(
+        ranked_transitions.c.new_state == state.name,
+        ranked_transitions.c.rank == 1,
+        ranked_transitions.c.label_name == labels.c.name,
+        labels.c.state_machine == state_machine.name,
+    ))
+
+    return [x for x, in conn.execute(active_participants)]
```

Comment:
ewww.

Reply:
Yeah, I couldn't find a better way to do this in the SQLA API.